packages/react/ds-app-launchpad/src/ui/FileTree/Provider.tsx

```diff
@@ -0,0 +1,127 @@
+import type React from "react";
+import { useCallback, useEffect, useRef } from "react";
+import Context from "./Context.js";
+import "./styles.css";
+import type { ManagedContextOptions, ProviderOptions } from "./types.js";
+
+const componentCssClassName = "ds file-tree";
+
+const Provider = ({
+  id,
+  className,
+  style,
+  children,
+  ...contextOptions
+}: ProviderOptions): React.ReactElement => {
+  // const focusedNodeRef = useRef<HTMLLIElement>(null);
+  const fileTreeRef = useRef<HTMLDivElement>(null);
+
+  const handleFocusNextNode: ManagedContextOptions["focusNextNode"] =
+    useCallback((direction) => {
+      if (!fileTreeRef.current) return;
+      const currentNode = fileTreeRef.current.querySelector<HTMLElement>(
+        "[data-path-hash]:focus",
+      );
+      const currentNodeHash = currentNode?.getAttribute("data-path-hash");
+      if (!currentNodeHash) return;
+
+      const nextNode =
+        direction === "up"
+          ? fileTreeRef.current.querySelector<HTMLElement>(
+              `[data-path-hash]:has(+ [data-path-hash="${currentNodeHash}"])`,
```

Comment:
`:has` is relatively new to browser baselines. It currently sits at ~93% support per [caniuse](https://caniuse.com/css-has). I wonder if there might be an implementation approach that has wider browser support that you could use, but ultimately this is a pretty high level of effort for this PR so I would suggest taking it up as a separate issue after this merges.